pkg/coscheduling/coscheduling.go

```diff
@@ -83,72 +94,110 @@ func New(_ *runtime.Unknown, handle framework.FrameworkHandle) (framework.Plugin
 // 2. Compare the timestamps of the initialization time of PodGroups.
 // 3. Compare the keys of PodGroups.
 func (cs *Coscheduling) Less(podInfo1 *framework.PodInfo, podInfo2 *framework.PodInfo) bool {
-	pod1 := podInfo1.Pod
-	pod2 := podInfo2.Pod
-	priority1 := pod.GetPodPriority(pod1)
-	priority2 := pod.GetPodPriority(pod2)
+	pgInfo1 := cs.getPodGroupInfo(podInfo1)
+	pgInfo2 := cs.getPodGroupInfo(podInfo2)
+
+	priority1 := pgInfo1.priority
+	priority2 := pgInfo2.priority
 
 	if priority1 != priority2 {
 		return priority1 > priority2
 	}
 
-	pgInfo1 := cs.getPodGroupInfo(podInfo1)
-	pgInfo2 := cs.getPodGroupInfo(podInfo2)
 	time1 := pgInfo1.timestamp
 	time2 := pgInfo2.timestamp
 
 	if !time1.Equal(time2) {
 		return time1.Before(time2)
 	}
 
-	key1 := fmt.Sprintf("%v/%v", podInfo1.Pod.Namespace, pgInfo1.name)
-	key2 := fmt.Sprintf("%v/%v", podInfo2.Pod.Namespace, pgInfo2.name)
-	return key1 < key2
+	return pgInfo1.key < pgInfo2.key
 }
 
-func (cs *Coscheduling) getPodGroupInfo(p *framework.PodInfo) *PodGroupInfo {
-	podGroupName, min, err := GetPodGroupLabels(p.Pod)
-	if err == nil && podGroupName != "" && min > 1 {
-		key := fmt.Sprintf("%v/%v", p.Pod.Namespace, podGroupName)
-		pgInfo, ok := cs.podGroupInfos.Load(key)
-		if !ok {
-			pgInfo = &PodGroupInfo{
-				name:      podGroupName,
-				timestamp: p.InitialAttemptTimestamp,
-			}
-			cs.podGroupInfos.Store(key, pgInfo)
+// getPodGroupkey returns a key of a PodGroup in the form of namespace/PodGroupName.
+func getPodGroupKey(namespace, podGroupName string) string {
+	return fmt.Sprintf("%v/%v", namespace, podGroupName)
+}
+
+// getPodGroupInfo creates a PodGroup if not exsiting;
+// and store it in a local cache map if the PodGroup has more than one Pod.
+func (cs *Coscheduling) getPodGroupInfo(podInfo *framework.PodInfo) *PodGroupInfo {
+	pod := podInfo.Pod
+	podGroupName, minAvailable, _ := GetPodGroupLabels(pod)
+	key := getPodGroupKey(pod.Namespace, podGroupName)
+
+	// If it is a PodGroup with more than one pod, check if there is a PodGroup in PodGroupInfos
+	if len(podGroupName) > 0 && minAvailable >= 1 {
+		pgInfo, exist := cs.podGroupInfos.Load(key)
+		if exist {
+			return pgInfo.(*PodGroupInfo)
 		}
-		return pgInfo.(*PodGroupInfo)
 	}
 
-	// If the pod is regular pod, return object of PodGroupInfo but not store in PodGroupInfos.
-	// The purpose is to facilitate unified comparison.
-	return &PodGroupInfo{name: "", timestamp: p.InitialAttemptTimestamp}
+	// If it's a non-existing PodGroup with more than 1 pod or a regular pod,
+	// create a PodGroup
+	pgInfo := &PodGroupInfo{
+		name:         podGroupName,
+		key:          key,
```

Comment:
Having two variables is mainly for extensibility. _key_ can be any unique ID, e.g., a randomly generated ID, GUID or UUID, etc. while podGroupName is a string defined in a pod spec or CRD. It looks redundant just because the current key is defined as _namespace/podgroupname_. Does it make sense?

Reply:
SG.